bundles/org.eclipse.e4.ui.workbench.compatibility.migration/src/org/eclipse/e4/ui/workbench/compatibiliy/migration/Converter.java

```diff
@@ -0,0 +1,933 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.e4.ui.model.application.MAddon;
+import org.eclipse.e4.ui.model.application.MApplication;
+import org.eclipse.e4.ui.model.application.MApplicationFactory;
+import org.eclipse.e4.ui.model.application.ui.MUIElement;
+import org.eclipse.e4.ui.model.application.ui.advanced.MAdvancedFactory;
+import org.eclipse.e4.ui.model.application.ui.advanced.MArea;
+import org.eclipse.e4.ui.model.application.ui.advanced.MPerspective;
+import org.eclipse.e4.ui.model.application.ui.advanced.MPerspectiveStack;
+import org.eclipse.e4.ui.model.application.ui.advanced.MPlaceholder;
+import org.eclipse.e4.ui.model.application.ui.basic.MBasicFactory;
+import org.eclipse.e4.ui.model.application.ui.basic.MPart;
+import org.eclipse.e4.ui.model.application.ui.basic.MPartSashContainer;
+import org.eclipse.e4.ui.model.application.ui.basic.MPartSashContainerElement;
+import org.eclipse.e4.ui.model.application.ui.basic.MPartStack;
+import org.eclipse.e4.ui.model.application.ui.basic.MStackElement;
+import org.eclipse.e4.ui.model.application.ui.basic.MTrimmedWindow;
+import org.eclipse.e4.ui.model.application.ui.basic.MWindow;
+import org.eclipse.e4.ui.workbench.IPresentationEngine;
+import org.eclipse.e4.ui.workbench.modeling.EPartService;
+import org.eclipse.e4.ui.workbench.persistence.common.CommonUtil;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.emf.ecore.resource.impl.ResourceSetImpl;
+import org.eclipse.swt.graphics.Rectangle;
+import org.eclipse.ui.IMemento;
+import org.eclipse.ui.WorkbenchException;
+import org.eclipse.ui.XMLMemento;
+import org.eclipse.ui.internal.e4.compatibility.ModeledPageLayoutUtils;
+
+/**
+ * The Implementation of the Converter interface.
+ *
+ * @since 3.3
+ *
+ */
+@SuppressWarnings("restriction")
+public class Converter {
+
+	/** Element id for the primary editor stack in an e4 application. */
+	private static final String PRIMARY_EDITOR_PART_STACK_ID = "org.eclipse.e4.primaryDataStack"; //$NON-NLS-1$
+
+	private static final String EDITOR_AREA_ID = "org.eclipse.ui.editorss"; //$NON-NLS-1$
+
+	private final Map<Object, Set<String>> visited = new LinkedHashMap<>();
+
+	/**
+	 * Constructor.
+	 */
+	public Converter() {
+	}
+
+	/**
+	 * Allows to convert an 3.x perspective to a serialized e4 applciation.
+	 *
+	 * @param sourcePath The path to the 3.x serialized perspective.
+	 * @param targetPath The path to the file to serialize the e4 application to.
+	 * @throws IOException           When reading/writing the files fails.
+	 * @throws FileNotFoundException When the files do not exist.
+	 * @throws WorkbenchException    When reading the 3.x perspective fails
+	 */
+	public static void convert(final String sourcePath, final String targetPath)

```

Comment:
`convert` is also a bit imprecise here, particularly with the relatively basic types being used. Something like `convertToE4Perspective` would be more explicit. Also if the arguments are Path, then let's use File or Path, not strings; and the arguments could tell `sourceMementoPath` and `targetMApplicationPath`